### PR TITLE
feat(app): add in-chat search with match navigation

### LINKED ISSUE #2230: Improve in-chat search discoverability and performance
## Summary

Users can search chats globally with Cmd+Shift+F today, but feedback indicates chat search is not discoverable enough and could be easier when a user wants to search only within the current chat/session.

## User feedback

> chats should be searchable

## Current behavior

- Cmd+Shift+F provides search, but users may not discover it.
- Search appears oriented around broader/global search rather than an obvious current-chat-only workflow.

## Proposed improvements

- Add a more visible in-chat search affordance in the session UI.
- Make it easy to scope search to the current chat/session.
- Improve performance for large/long chats.
- Consider highlighting matches and providing next/previous navigation within the current chat.

## Notes

This came from OpenWork app feedback on desktop 0.16.4.

### PR DESCRIPTION
**Pull request by nehaprasad-dev**
#### Summary
- Add in-chat search for the current session with a header search button and find bar
- Highlight matches and move between them with prev/next controls

#### Why
- Chat search was hard to find (Cmd+Shift+F only) and felt global, not session-scoped

#### Issue
Closes #2230

#### Scope
- Current-chat search UI, Cmd/Ctrl+F, match highlighting, next/previous navigation, debounced search

#### Out of scope
- Changes to global session search (Cmd/Ctrl+Shift+F)

#### Testing
- Ran
pnpm typecheck in apps/app
- Result
pass/fail: pass
if fail, exact files/errors: N/A

#### CI status
- pass: not run here
- code-related failures: none known
- external/env/auth blockers: bun not available locally for unit tests

#### Manual verification
- Open a chat and click the search icon in the header
- Type a term and confirm matches are highlighted
- Use ↑↓ or Enter to move between matches; Escape to close

### CODE REVIEW

File: apps/app/src/react-app/domains/session/search/use-session-chat-search.ts [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+      if (event.key === "Enter") {
+        event.preventDefault();
+        if (event.shiftKey) {
+          goPrev();
+        } else {
+          goNext();
+        }
+      }
+    };
+    window.addEventListener("keydown", handleKeyDown);
```
[the remark this answers is not in the document]
Comment (Muhammad-Owais-Warsi) [reply apps/app/src/react-app/domains/session/search/use-session-chat-search.ts:115]: The only suitable way I see here is to, check the targets before executing the keywords to prevent the blocking native behavior in other interactive elements.

a basic example

```js
if(event.target === "SEARCH") {
...
}
```

**Comment by Abhijeet1005** [comment]
Took this for a local spin on Windows 11 (Node 24.13, pnpm 11.4, Bun 1.3.14) to help move it along. Nice feature — it does match the UX in #2230 (scoped search, highlight, next/prev, debounced). Notes below.

**Verified locally** (the repo's test/i18n CI workflows haven't run here yet — fork PRs need maintainer approval):
- `pnpm --filter @openwork/app typecheck` → pass
- `bun test apps/app/tests/session-chat-search.test.ts` → 3 pass / 0 fail

**1. Confirming the `Enter` handler issue** that cubic (and @Muhammad-Owais-Warsi) flagged at `use-session-chat-search.ts:115`. The keydown listener is on `window` while the bar is open, so once focus moves back to the composer, `Enter` (send) and `Shift+Enter` (newline) get `preventDefault`'d into match-navigation. Scoping the handler to the search input (or gating on `document.activeElement === inputRef.current`) fixes it.

**2. Match count and highlights are computed from two different sources, so they desync.** `countChatSearchMatches` counts only `type:"text"` message parts, but `applyTextHighlights` walks the rendered DOM and rejects `pre, code` (`text-highlights.ts:42`). Repro: search a term that also appears inside a code block — that occurrence is counted but never highlighted, so the "N of M" indicator overstates the findable matches. Stepping `next`/`prev` onto the phantom index then clears the active highlight (no `<mark>` gets `data-search-active`) and snaps back to the first match instead of advancing. It diverges the other way too — matches in tool output / reasoning get highlighted but aren't counted. Worth picking one source of truth (e.g. count from the same DOM walk the highlighter uses) so the indicator and navigation stay consistent — most visible in long chats with code, which is the #2230 use case.

**3. (minor)** The header button hardcodes `title="Search this chat (⌘F)"`, but the shortcut handler correctly uses `Ctrl` on Windows/Linux — so the tooltip shows the wrong key off-Mac.

**4. (minor)** The user-facing strings ("Search this chat…", "Previous/Next match", "No matches", "N of M") are hardcoded rather than going through `t()` / `i18n/locales/en.ts`. Won't fail CI, but they won't localize to the other 9 languages.

Happy to send a small follow-up patch for 1 & 2 if that's useful.

**Comment by Abhijeet1005** [comment]
Drafted the follow-up for **1 & 2** since it was quick — feel free to fold it in or tweak. Verified on the same setup: `pnpm --filter @openwork/app typecheck` passes and the full `bun test tests/` suite is green (68/0).

- **1 — scope `Enter` to the focused search field**, so the composer's send / Shift+Enter keep working while the bar is open.
- **2 — one source of truth for the count**: `applyTextHighlights` now returns the number of marks it creates, and the hook drives `matchCount` from that same DOM pass — so "N of M" and next/prev always agree with what's actually highlighted (fixes the code-block case where the count and highlights diverged).

```diff
diff --git a/apps/app/src/react-app/domains/session/search/use-session-chat-search.ts b/apps/app/src/react-app/domains/session/search/use-session-chat-search.ts
--- a/apps/app/src/react-app/domains/session/search/use-session-chat-search.ts
+++ b/apps/app/src/react-app/domains/session/search/use-session-chat-search.ts
@@ -1,4 +1,4 @@
-import { useCallback, useEffect, useMemo, useRef, useState } from "react";
+import { useCallback, useEffect, useRef, useState } from "react";
 import type { UIMessage } from "ai";
 
 import {
@@ -6,7 +6,6 @@ import {
   clearTextHighlights,
   scrollActiveMatchIntoView,
 } from "@/react-app/domains/session/surface/text-highlights";
-import { countChatSearchMatches } from "./session-chat-search";
 
 const DEBOUNCE_MS = 200;
 
@@ -35,14 +34,13 @@ export function useSessionChatSearch(input: UseSessionChatSearchInput) {
   const { messages, contentRef, open, onOpenChange, enableShortcuts = true } = input;
   const [query, setQuery] = useState("");
   const [activeIndex, setActiveIndex] = useState(0);
+  // Match count is measured from the same DOM pass that draws the highlights
+  // (see the highlight effect below), so the "N of M" indicator and next/prev
+  // navigation can never disagree with what is actually highlighted.
+  const [matchCount, setMatchCount] = useState(0);
   const inputRef = useRef<HTMLInputElement>(null);
   const debouncedQuery = useDebouncedValue(query.trim(), DEBOUNCE_MS);
 
-  const matchCount = useMemo(
-    () => countChatSearchMatches(messages, debouncedQuery),
-    [messages, debouncedQuery],
-  );
-
   const safeActiveIndex = matchCount > 0 ? activeIndex % matchCount : 0;
 
   useEffect(() => {
@@ -53,6 +51,7 @@ export function useSessionChatSearch(input: UseSessionChatSearchInput) {
     if (!open) {
       setQuery("");
       setActiveIndex(0);
+      setMatchCount(0);
       const root = contentRef.current;
       if (root) {
         clearTextHighlights(root);
@@ -73,7 +72,8 @@ export function useSessionChatSearch(input: UseSessionChatSearchInput) {
     }
 
     const frame = window.requestAnimationFrame(() => {
-      applyTextHighlights(root, debouncedQuery, safeActiveIndex);
+      const count = applyTextHighlights(root, debouncedQuery, safeActiveIndex);
+      setMatchCount(count);
       scrollActiveMatchIntoView(root);
     });
     return () => window.cancelAnimationFrame(frame);
@@ -103,7 +103,9 @@ export function useSessionChatSearch(input: UseSessionChatSearchInput) {
         onOpenChange(false);
         return;
       }
-      if (event.key === "Enter") {
+      // Only intercept Enter while the search field itself is focused, so the
+      // composer's send / Shift+Enter newline keep working while the bar is open.
+      if (event.key === "Enter" && document.activeElement === inputRef.current) {
         event.preventDefault();
         if (event.shiftKey) {
           goPrev();
diff --git a/apps/app/src/react-app/domains/session/surface/text-highlights.ts b/apps/app/src/react-app/domains/session/surface/text-highlights.ts
--- a/apps/app/src/react-app/domains/session/surface/text-highlights.ts
+++ b/apps/app/src/react-app/domains/session/surface/text-highlights.ts
@@ -18,7 +18,7 @@ export function scrollActiveMatchIntoView(root: HTMLElement) {
   target?.scrollIntoView({ block: "center", behavior: "smooth" });
 }
 
-export function applyTextHighlights(root: HTMLElement, query: string, activeIndex = 0) {
+export function applyTextHighlights(root: HTMLElement, query: string, activeIndex = 0): number {
   const needle = query.trim().toLowerCase();
   // Fast path: if search is inactive, avoid walking large message DOM trees.
@@ -27,7 +27,7 @@
     if (root.querySelector(`mark[${SEARCH_HIGHLIGHT_MARK_ATTR}="true"]`)) {
       clearTextHighlights(root);
     }
-    return;
+    return 0;
   }
@@ -88,4 +88,6 @@
     node.parentNode?.replaceChild(fragment, node);
   });
+
+  return matchCounter;
 }
```

That leaves `countChatSearchMatches` unused, so I also removed `session-chat-search.ts` + its test (the count now comes from the highlight pass). Net `+15 / −72` overall.

Happy to push this as a branch or open a PR against `feat/cht-srch-dis-perf` if that's easier than applying by hand.

**Comment by Pablosinyores** [comment]
Really like the shape here — a pure `countChatSearchMatches` with its own test, debounced query, memoized count.

One thing worth double-checking: the match count and the highlights come from two independent passes. `countChatSearchMatches` searches `message.parts.filter(text).map(text).join("")` (concatenated with no separator), while `applyTextHighlights` walks the rendered DOM and counts its own `matchCounter`. Two risks from the `join("")`: (1) a query can match across the seam of two adjacent text parts, giving a count with no contiguous on-screen match; (2) if the DOM highlighter matches per text node, those seam matches aren't highlightable, so the "X of N" indicator and next/prev navigation can drift out of sync with what's actually shown. Might be worth joining parts with a separator (e.g. `\n`) and/or asserting count parity against the rendered `mark`s in the test for a multi-part message. @nehaprasad-dev

**Comment by Abhijeet1005** [comment]
@Pablosinyores good catch, and yes — this is real. `messageSearchText` joins parts with `join("")` (`session-chat-search.ts:7`) while the highlighter walks per text-node (it accepts a node only if that single node `.includes(needle)`), so a query straddling two parts — e.g. text split around a tool-call part — is counted but never highlightable. And since `matchCount` is driven by that data pass (`use-session-chat-search.ts:41`), the "N of M" indicator and next/prev key off a number the marks don't match.

This overlaps with the follow-up I drafted above (issue 2): have `applyTextHighlights` return its mark count and drive `matchCount` off that one DOM pass — then count == marks by construction, and the cross-seam case stops being counted (correctly, since it can't be shown anyway).

Trade-off worth a maintainer call: your `join("\n")` keeps the nice pure `countChatSearchMatches` + its DOM-free unit test but leaves two passes to keep in sync; collapsing to the DOM pass removes the second pass entirely but drops that unit test. Worth noting the parity assertion you suggested (count == rendered `mark`s) needs a DOM environment — the app's `bun:test` setup has none today, the existing test is pure — so it's a small happy-dom/jsdom add either way.
